tangerine/installation.cpp

```diff
@@ -17,20 +17,39 @@
 
 #if !_WIN64
 #include <unistd.h>
+#include <sys/types.h>
+#include <pwd.h>
+#include <iostream>
+#include <cerrno>
+#include <cstring>
 #endif
 
+namespace fs = std::filesystem;
+
+#if !_WIN64
+const char *const tangerine_app_id =
+	// TODO: consider a reverse-dns name with the escaping recommended in:
+	// https://docs.gtk.org/gio/type_func.Application.id_is_valid.html
+	// See rationale in: https://docs.gtk.org/gtk4/migrating-3to4.html#set-a-proper-application-id
+	"tangerine";
+std::optional<fs::path> tangerine_get_xdg_state_home();
+#endif
 
 TangerinePaths::TangerinePaths(int argc, char* argv[])
 {
+	fs::path ExecutablePath;
 #if _WIN64
 	// TODO: This is fine for standalone builds, but we will want to do something
 	// else for future library builds.  Maybe GetModuleFileNameW?
-	ExecutablePath = std::filesystem::absolute(argv[0]);
+	ExecutablePath = fs::absolute(argv[0]);
 
 #elif EMBED_RACKET
-	ExecutablePath = std::filesystem::path(racket_get_self_exe_path(argv[0]));
+	// On Windows, `racket_get_self_exe_path()` returns UTF-8.
+	ExecutablePath = fs::path(racket_get_self_exe_path(argv[0]));
```

Comment:
I meant to ask before: why do Racket-enabled linux builds have special behavior here?  `ExecutablePath` should be the path to the Tangerine executable.

Reply:
Yes, `ExecutablePath` should be the path to the Tangerine executable. Racket happens to have code to find that path which is portable, deals with various edge cases, and is not maintained by me. Recently (in https://github.com/racket/racket/issues/4483) it got some enhancements from LLVM and became part of the public C API, because it can be useful for finding arguments to `racket_boot` (that's basically why it exists). The `/proc/self/exe` strategy is simple and probably good enough (I'm not sure many people will start Tangerine in chroots, and I don't think you have any intention of supporting GNU/Hurd or the BSDs, let alone stranger systems), but the off-the-shelf solution seemed robust enough to be worth using when it's available anyway.

Reply:
It's probably fine for now then, but I would like to eventually replace it want a more robust solution that isn't an optional dependency in the future.  Also keep in mind that once Tangerine can be used as a full featured library (and thus wrapped as a nice Racket library), there are few good reasons to keep embedding Racket in Tangerine.  But that is a ways off anyway, and this works right now at least.

Running Tangerine from a chroot seems like a reasonable case to expect, and we should try to support it well.

As for Hurd, BSD, and Death Station 9000 - I don't anticipate these being significant targets worth investing substantial time on.  However, if they are trivial to port to, well, maybe I guess?  The two concerns that come to mind right now are 1) I have tentative plans to replace OpenGL with Vulkan, which some of these platforms might not have support for (BSD didn't last time I checked), and 2) how regression testing would work.  Right now regression testing is easy - run the build on Windows, and then I run it on WSL w/ guix.  If you are reasonably sure that a working Linux build implies a working Hurd build with only some minor changes limited to the installation sources, then it might be fine to have unofficial support for it?

Reply:
I think I'll probably integrate this library to determine the exe path on all platforms: https://github.com/gpakosz/whereami

Reply:
Ok the new library has been integrated as of 24cfebffdc499982ae383aabbeb253d17d5e2c29 so this is no longer an issue.  We might need to mod that library for more exotic build targets, but it looks like it supports substantially more than Tangerine does.  I've confirmed that this works on Linux and Windows, but obviously let me know if there are any problems.

As for the chroot case, it looks like a missing proc file system is something that can be resolved by the person setting up the chroot, so I don't think handling that case is something Tangerine should try to do.  Instead, I made it so it prints out a hint if it fails to determine the exe path and can't find `/proc` on `!_WIN64`.  If the path can't be found via the library, Tangerine now just prints an error and terminates.  Seems like the safest thing to do here.

Reply:
> As for the chroot case, it looks like a missing proc file system is something that can be resolved by the person setting up the chroot, so I don't think handling that case is something Tangerine should try to do. Instead, I made it so it prints out a hint if it fails to determine the exe path and can't find `/proc` on `!_WIN64`. If the path can't be found via the library, Tangerine now just prints an error and terminates. Seems like the safest thing to do here.

This makes sense to me. Probably I've been overthinking this. Until https://github.com/racket/racket/commit/5983026fe3cad6616dedb79dd4970d1b5aa1c6f2 (December), Racket's fallback was `exit(1)`. A Linux chroot *without `/proc` mounted* is a very strange and restricted environment: probably it's worthwhile for a programming language's runtime system to make heroic efforts to work in all sorts of environments, but users can't expect arbitrary applications to work without `/proc`.

Some kind of library function seems like a good thing. It might be worth revisiting when Tangerine-as-a-library becomes a thing, too: if at that point the library just needs shaders, maybe it would make sense to embed them in the library itself? IDK.